drivers/frequency/adf4382/iio_adf4382.c

```diff
@@ -0,0 +1,938 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <stdlib.h>
+#include <stdio.h>
+#include <inttypes.h>
+#include "no_os_error.h"
+#include "no_os_util.h"
+#include "iio_adf4382.h"
+#include "adf4382.h"
+#include "no_os_alloc.h"
+
+/**
+ * @brief Supported charge pump currents
+ */
+static const int ci_table[16][2] = {
+	{0, 700000},
+	{0, 900000},
+	{1, 100000},
+	{1, 300000},
+	{1, 500000},
+	{1, 900000},
+	{2, 300000},
+	{2, 700000},
+	{3, 100000},
+	{3, 900000},
+	{4, 700000},
+	{5, 500000},
+	{6, 300000},
+	{7, 900000},
+	{9, 500000},
+	{11, 100000}
+};
+
+/**
+ * @brief Helper function to read a pair of values in a table with 2 elements.
+ * @param tb	- Input table.
+ * @param size	- Size of the table.
+ * @param val1	- The first value to be found.
+ * @param val2	- The second value to be found.
+ * @return i	- The index of the pair if found or an error if there is no
+ * 		  pair of values in the table.
+ */
+static int adf4382_iio_find_2d_row(const int (*tbl)[2], const int size,
+				   const int val1, const int val2)
+{
+	int i;
+
+	for (i = 0; i < size; i++) {
+		if (tbl[i][0] == val1 && tbl[i][1] == val2)
+			return i;
+	}
+
+	return -EINVAL;
+}
+
+/**
+ * @brief Wrapper for reading adf4382 register.
+ * @param dev	  - The iio device structure.
+ * @param reg	  - Address of the register to be read from.
+ * @param readval - Read data.
+ * @return	  - Result of the reading procedure.
+ */
+static int adf4382_iio_read_reg(struct adf4382_iio_dev *dev, uint32_t reg,
+				uint32_t *readval)
+{
+	return adf4382_spi_read(dev->adf4382_dev, (uint16_t)reg,
+				(uint8_t *)readval);
+}
+
+/**
+ * @brief Wrapper for writing adf4382 register.
+ * @param dev	   - The iio device structure.
+ * @param reg	   - Address of the register to be written to.
+ * @param writeval - Data to be written.
+ * @return	   - Result of the writing procedure.
+ */
+static int adf4382_iio_write_reg(struct adf4382_iio_dev *dev, uint32_t reg,
+				 uint32_t writeval)
+{
+	return adf4382_spi_write(dev->adf4382_dev, (uint16_t)reg,
+				 (uint8_t)writeval);
+}
+
+/**
+ * @brief Handles the read request for frequency attribute.
+ * @param dev     - The iio device structure.
+ * @param buf	  - Command buffer to be filled with requested data.
+ * @param len     - Length of the received command buffer in bytes.
+ * @param channel - Command channel info.
+ * @param priv    - Command attribute id.
+ * @return 	  - The size of the read data in case of success, error code
+ * 		    otherwise.
+ */
+static int adf4382_iio_read_freq(void *dev, char *buf, uint32_t len,
+				 const struct iio_ch_info *channel,
+				 intptr_t priv)
+{
+	struct adf4382_iio_dev *iio_adf4382 = (struct adf4382_iio_dev *)dev;
+	struct adf4382_dev *adf4382 = iio_adf4382->adf4382_dev;
```

Comment:
this should be done after the null check of iio_adf4382, please see all occurrences in the file